Duplicati.Library.Backend.Storx/StorxBackend.cs

```diff
@@ -0,0 +1,491 @@
+        }
+
+        public void Test()
+        {
+            this.TestList();
+        }
+
+        public void CreateFolder()
+        {
+           // GetFolder(1);
+        }
+
+        #endregion
+
+        #region IDisposable Members
+
+        public void Dispose()
+        {
+        }
+
+        #endregion
+
+
+        public string[] DNSName
+        {
+            get { return new string[] { new System.Uri(root).Host }; }
+        }
+
+        public void Get(string remotename, System.IO.Stream stream)
+        {
+            StorxFileList fl1 = GetFiles();
+            if (fl1.Files != null)
+            {
+                var fileID = fl1.Files.FindAll(x => (x.name + "." + x.type) == remotename)[0];
+                string endpoint = "storage/file/" + fileID.fileId;
+                var req = CreateRequest(endpoint);
+                req.Method = System.Net.WebRequestMethods.Http.Get;
+                var rreq = new Utility.AsyncHttpRequest(req);
+                using (var s = rreq.GetResponseStream())
+                    Library.Utility.Utility.CopyStream(s, stream);
+            }
+
+        }
+
+        public Task PutAsync(string remotename, System.IO.Stream stream, CancellationToken cancelToken)
+        {
+            string endpoint = root+"storage/folder/" + m_root_folder + "/upload";
+            try
+            {
+                byte[] formData = ReadToEnd(stream);
+                var webClient = new WebClient();
+                string boundary = "------------------------" + DateTime.Now.Ticks.ToString("x");
+                string contentType = "multipart/form-data; boundary=" + boundary;
+                webClient.Headers.Add("Content-Type", "multipart/form-data; boundary=" + boundary);
+                webClient.Headers.Add("Authorization", "Bearer " + authorization);
+                webClient.Headers.Add("storx-mnemonic", "scissors huge model dragon mind distance spider fit cute gym rain now dragon solution online monkey describe glow gossip cancel elder misery pony mask");
+                var fileData = webClient.Encoding.GetString(formData);
+                var package = string.Format("--{0}\r\nContent-Disposition: form-data; name=\"xfile\"; filename=\"{1}\"\r\nContent-Type: {2}\r\n\r\n{3}\r\n--{0}--\r\n", boundary, remotename, contentType, fileData);
+                var nfile = webClient.Encoding.GetBytes(package);
+                byte[] resp = webClient.UploadData(endpoint, "POST", nfile);
+                return Task.FromResult(true);
+            }
+
+            catch (System.Net.WebException wex)
+            {
+                if (wex.Response is HttpWebResponse response && response.StatusCode == System.Net.HttpStatusCode.NotFound)
+                    throw new FileMissingException(wex);
+                else
+                    throw new Exception(getResponseBodyOnError(endpoint, wex));
+            }
+            
+        }
+
+        public static byte[] ReadToEnd(System.IO.Stream stream)
```

Comment:
This function is also in the utility project

Reply:
all changes are done, kindly do the needful